test/modern_bpf/test_suites/syscall_exit_suite/open_by_handle_at_x.cpp

```diff
@@ -0,0 +1,215 @@
+
+TEST(SyscallExit, open_by_handle_atX_success)
+{
+	auto evt_test = new event_test(__NR_open_by_handle_at, EXIT_EVENT);
+
+	evt_test->enable_capture();
+
+	/*=============================== TRIGGER SYSCALL  ===========================*/
+
+	/*
+	 * 1. Open a temp file.
+	 */
+	int dirfd = AT_FDCWD;
+	const char *pathname = ".";
+	int flags = O_RDWR | O_TMPFILE | O_DIRECTORY;
+	mode_t mode = 0;
+	int fd = syscall(__NR_openat, dirfd, pathname, flags, mode);
+	assert_syscall_state(SYSCALL_SUCCESS, "openat", fd, NOT_EQUAL, -1);
+
+	/* Allocate file_handle structure. */
+	struct file_handle *fhp;
+	int fhsize = sizeof(*fhp);
+	fhp = (struct file_handle *)malloc(fhsize);
+	if(fhp == NULL)
+	{
+		FAIL() << "Error in allocating the `struct file_handle` with malloc" << std::endl;
+	}
+
+	/* Make an initial call to name_to_handle_at() to discover the size required for the file handle.
+	 * The caller can discover the required size for the file_handle structure by making a call in which handle->handle_bytes is zero;
+	 * in this case, the call fails with the error EOVERFLOW and handle->handle_bytes is set to indicate the required size;
+	 */
+	int mount_id;
+	flags = 0;
+	fhp->handle_bytes = 0;
+	assert_syscall_state(SYSCALL_FAILURE, "name_to_handle_at", syscall(__NR_name_to_handle_at, dirfd, pathname, fhp, &mount_id, flags));
+
+	/*
+	 * 2. Reallocate file_handle structure with the correct size.
+	 */
+	fhsize = sizeof(*fhp) + fhp->handle_bytes;
+	fhp = (struct file_handle *)realloc(fhp, fhsize);
+	if(fhp == NULL)
+	{
+		FAIL() << "Error in allocating the `struct file_handle` with realloc" << std::endl;
+	}
+
+	/*
+	 * 3. Get file handle.
+	 */
+	assert_syscall_state(SYSCALL_SUCCESS, "name_to_handle_at", syscall(__NR_name_to_handle_at, dirfd, pathname, fhp, &mount_id, flags), NOT_EQUAL, -1);
+
+	/*
+	 * 4. Retrieve the `mount_fd`.
+	 */
+	char *linep = NULL;
+	size_t lsize = 0;
+	char mount_path[4096];
+	int mi_mount_id = 0;
+	int found = 0;
+	ssize_t nread = 0;
+	FILE *fp = NULL;
+
+	fp = fopen("/proc/self/mountinfo", "r");
+	if(fp == NULL)
+	{
+		FAIL() << "Error in opening /proc/self/mountinfo" << std::endl;
+	}
+
+	while(!found)
+	{
+		nread = getline(&linep, &lsize, fp);
+		if(nread == -1)
+		{
+			break;
+		}
+
+		nread = sscanf(linep, "%d %*d %*s %*s %s", &mi_mount_id, mount_path);
+		if(nread != 2)
+		{
+			FAIL() << "Bad sscanf()" << std::endl;
+		}
+
+		if(mi_mount_id == mount_id)
+		{
+			found = 1;
+		}
+	}
+
+	if(!found)
+	{
+		FAIL() << "Could not find mount point" << std::endl;
+	}
+
+	/*
+	 * 5. Call `open_by_handle_at`.
+	 */
+	int mount_fd = open(mount_path, O_RDONLY);
+	flags = O_RDONLY;
+	int open_by_handle_fd = syscall(__NR_open_by_handle_at, mount_fd, fhp, flags);
+	assert_syscall_state(SYSCALL_SUCCESS, "open_by_handle_at", open_by_handle_fd, NOT_EQUAL, -1);
+
+	/*
+	 * 6. Get the current working directory.
+	 */
+	char tmp[4096];
+	char* err = getcwd(tmp, 4096);
```

Comment:
Hi @Andreagit97,

with the discussed changed to the bpf_probe_read, I see a test case failure here:
````
[ RUN      ] SyscallExit.open_by_handle_atX_success
/root/git/falcosecurity/libs/test/modern_bpf/event_class/event_class.cpp:265: Failure
Expected equality of these values:
  size
    Which is: 30
  expected_size
    Which is: 35
>>>>> length of the param is not correct. Param id = 4

/root/git/falcosecurity/libs/test/modern_bpf/event_class/event_class.cpp:240: Failure
Expected equality of these values:
  m_event_params[m_current_param].valptr
    Which is: "/git/falcosecurity/libs/build"
  param
    Which is: "/root/git/falcosecurity/libs/build"
>>>>> value of the param is not correct. Param id = 4

[  FAILED  ] SyscallExit.open_by_handle_atX_success (0 ms)
````
The syscall reports the path w/o the `/root` prefix. Please note that `/root` is a mount point on my dev system. Needs the result of open_by_handle_fd be considered relative to the mount_fd in which case, the path to which `mount_fd` points needs to be removed from `tmp`?

Reply:
Uhm this is quite strange...Just a few points:
* I didn't understand if you face the problem only when you replace `bpf_probe_read_str()` with `bpf_probe_read_kernel_srt()` or always...
* Looking at how it is computed kernel side I expect to have an absolute path with also the `/root` component :thinking: could you please print the `mount_path` just to understand what is going on? as a second attempt you can replace the `mount_fd` in the `open_by_handle_at` call with `AT_FDCWD` obtaining something like

```c
int open_by_handle_fd = syscall(__NR_open_by_handle_at, AT_FDCWD, fhp, flags);
```

Reply:
Uhm reading your question again, if `/root` is a mount point probably it's correct to have a path like `/git/falcosecurity/libs/build`, probably we need to change the userspace part in order to get the right it works :thinking:

Reply:
Hi @Andreagit97,

>     * I didn't understand if you face the problem only when you replace `bpf_probe_read_str()` with `bpf_probe_read_kernel_srt()` or always...

I had to replace with bpf_probe_read_kernel_str() to obtain the proper path pointers (if I use use the user variant then nothing is returned). Basically, I added a _user and _kernel variant for the push__charbuf along with this change:

```diff
--- a/driver/modern_bpf/helpers/store/auxmap_store_params.h
+++ b/driver/modern_bpf/helpers/store/auxmap_store_params.h
@@ -367,7 +367,7 @@ static __always_inline void auxmap__store_path_from_fd(struct auxiliary_map *aux
        {
                if(path_pointers[k])
                {
-                       total_size += push__charbuf(auxmap->data, &auxmap->payload_pos, path_pointers[k], MAX_PARAM_SIZE);
+                       total_size += push__charbuf_kernel(auxmap->data, &auxmap->payload_pos, path_pointers[k], MAX_PARAM_SIZE);
                        push__previous_character(auxmap->data, &auxmap->payload_pos, '/');
                }
        }
```

> 
>     * Looking at how it is computed kernel side I expect to have an absolute path with also the `/root` component thinking could you please print the `mount_path` just to understand what is going on? as a second attempt you can replace the `mount_fd` in the `open_by_handle_at` call with `AT_FDCWD` obtaining something like

Basically, there is no change when using the AT_FDCWD value. Further, the passes when calling from an CWD mounted below / (no separate device). For example:
````
# (cd /home && /root/git/falcosecurity/libs/build/test/modern_bpf/bpf_test --gtest_filter=SyscallExit.open_by_handle_atX_success)
[...]
Note: Google Test filter = SyscallExit.open_by_handle_atX_success
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from SyscallExit
[ RUN      ] SyscallExit.open_by_handle_atX_success
XXX mountpath = '/'
XXX readlink: /home
[       OK ] SyscallExit.open_by_handle_atX_success (0 ms)
[----------] 1 test from SyscallExit (0 ms total)
````
I added some debug output ("XXX") displaying the `mount_path` as well as the result of readlink `"/proc/self/fd/%i", open_by_handle_fd`. Running this again on directory with a different mount point with /, e.g., /tmp (typically a tmpfs), I see the same failure:
````
# (cd /tmp && /root/git/falcosecurity/libs/build/test/modern_bpf/bpf_test --gtest_filter=SyscallExit.open_by_handle_atX_s
uccess)
[...]
Note: Google Test filter = SyscallExit.open_by_handle_atX_success
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from SyscallExit
[ RUN      ] SyscallExit.open_by_handle_atX_success
XXX mountpath = '/tmp'
XXX readlink: /tmp
/root/git/falcosecurity/libs/test/modern_bpf/event_class/event_class.cpp:265: Failure
Expected equality of these values:
  size
    Which is: 2
  expected_size
    Which is: 5
>>>>> length of the param is not correct. Param id = 4

/root/git/falcosecurity/libs/test/modern_bpf/event_class/event_class.cpp:240: Failure
Expected equality of these values:
  m_event_params[m_current_param].valptr
    Which is: "//"
  param
    Which is: "/tmp"
>>>>> value of the param is not correct. Param id = 4

[  FAILED  ] SyscallExit.open_by_handle_atX_success (0 ms)
````

Also I have added with this change:
````diff
+++ b/driver/modern_bpf/helpers/store/auxmap_store_params.h
@@ -322,6 +322,7 @@ static __always_inline void auxmap__store_path_from_fd(struct auxiliary_map *aux
                }
                path_components++;
                BPF_CORE_READ_INTO(&path_pointers[k], de, d_name.name);
+               bpf_printk("XXX: %s\n", path_pointers[k]);
                de_p = de;
                BPF_CORE_READ_INTO(&de, de, d_parent);
        }
```
The following is printed to the trace log (empty lines removed here):
````
[root@m8345024 build]# cat /sys/kernel/debug/tracing/trace
# tracer: nop
#
# entries-in-buffer/entries-written: 3/3   #P:2
#
#                                _-----=> irqs-off/BH-disabled
#                               / _----=> need-resched
#                              | / _---=> hardirq/softirq
#                              || / _--=> preempt-depth
#                              ||| / _-=> migrate-disable
#                              |||| /     delay
#           TASK-PID     CPU#  |||||  TIMESTAMP  FUNCTION
#              | |         |   |||||     |         |
        bpf_test-380152  [001] d.... 693203.143029: bpf_trace_printk: XXX: home
        bpf_test-380152  [001] d.... 693203.143032: bpf_trace_printk: XXX: /
        bpf_test-380154  [000] d.... 693215.193032: bpf_trace_printk: XXX: /
````

For references on the original failure:
````
        bpf_test-379619  [000] d.... 690825.593151: bpf_trace_printk: XXX: build
        bpf_test-379619  [000] d.... 690825.593153: bpf_trace_printk: XXX: libs
        bpf_test-379619  [000] d.... 690825.593154: bpf_trace_printk: XXX: falcosecurity
        bpf_test-379619  [000] d.... 690825.593154: bpf_trace_printk: XXX: git
        bpf_test-379619  [000] d.... 690825.593155: bpf_trace_printk: XXX: /
````

Hope this information helps.

Reply:
Well, I guess in addition to
```c
static __always_inline void auxmap__store_path_from_fd(struct auxiliary_map *auxmap, s32 fd)
{       

        u16 total_size = 0;
        u8 path_components = 0;
        unsigned long path_pointers[MAX_PATH_POINTERS] = {0};
        struct file *f = extract__file_struct_from_fd(fd);
        struct dentry *de = BPF_CORE_READ(f, f_path.dentry);
```
and
````
// from include/linux/path.h
struct path {
        struct vfsmount *mnt;
        struct dentry *dentry;
} __randomize_layout; 

// and from include/linux/mount.h
struct vfsmount {
        struct dentry *mnt_root;        /* root of the mounted tree */
        struct super_block *mnt_sb;     /* pointer to superblock */
        int mnt_flags;
        struct user_namespace *mnt_userns;
} __randomize_layout;

```
I think the mnt->mnt_root needs to considered as well?

Reply:
Yeah, it seems clear that we have a problem with mount points different from `/` :thinking: Thank you for all the debug information they are very precious!! I will take a look to if we can find a simple way to obtain the right path according to the mount point, but this seems to be a tricky problem for BPF this is why they introduced the `bpf_d_path` helper, but unfortunately, we cannot use it :( Let's see if we are able to solve the problem otherwise, I would open an issue to track it, and probably I would go on since I see this as a non-blocking feature but as a nice to have, let me know what you think about it :) 

And BTW you helped me in spotting a but that I've fixed, it should be `/` and not `//`, so thank you very much for this!!

```
/root/git/falcosecurity/libs/test/modern_bpf/event_class/event_class.cpp:240: Failure
Expected equality of these values:
  m_event_params[m_current_param].valptr
    Which is: "//" <--- this was a bug, I've solved it
  param
    Which is: "/tmp"
>>>>> value of the param is not correct. Param id = 4
```

Reply:
>  Let's see if we are able to solve the problem otherwise, I would open an issue to track it, and probably I would go on since I see this as a non-blocking feature but as a nice to have, let me know what you think about it :)

I am fine with tracking it separately... maybe a short comment or reference to the issue in the test case about the mount point to let users know about failures.

> And BTW you helped me in spotting a but that I've fixed, it should be `/` and not `//`, so thank you very much for this!!

 you are welcome!

Reply:
I've tried different places to catch the real root:
* https://elixir.bootlin.com/linux/latest/source/include/linux/fs.h#L1447
* or something like `BPF_CORE_READ_INTO(&path_pointers[0], f, f_path.mnt, mnt_root, d_name.name);` 

but nothing to do it always returns `/` instead of the mount point root like `/root` or `/tmp`

Reply:
IMHO this is a different issue and should be addressed in a separate PR :) (possibly also for kmod and old ebpf right?)
Btw thanks for all the debug informations you provided @hbrueckner !

Reply:
kmod works fine since it uses the `d_path` kernel function, we have issues only with the 2 BPF probes :(